legion/engine/llri-vk/source.cpp

```diff
@@ -1,14 +1,259 @@
+
+        void* pNext = nullptr;
+
+        //Variables that need to be stored outside of scope
+        vk::ValidationFeaturesEXT features; 
+        std::vector<vk::ValidationFeatureEnableEXT> enables;
+
+        for (uint32_t i = 0; i < desc.numExtensions; i++)
+        {
+            auto& extension = desc.extensions[i];
+            switch(extension.type)
+            {
+                case instance_extension_type::APIValidation:
+                {
+                    if (extension.apiValidation.enable)
+                        layers.push_back("VK_LAYER_KHRONOS_validation");
+                    break;
+                }
+                case instance_extension_type::GPUValidation:
+                {
+                    if (extension.gpuValidation.enable)
+                    {
+                        enables = {
+                            vk::ValidationFeatureEnableEXT::eGpuAssisted,
+                            vk::ValidationFeatureEnableEXT::eGpuAssistedReserveBindingSlot
+                        };
+                        features = vk::ValidationFeaturesEXT(enables, {});
+                        features.pNext = pNext; //Always apply pNext backwards to simplify optional chaining
+                        pNext = &features;
+                    }
+                    break;
+                }
+                default:
+                {
+                    llri::destroyInstance(result);
+                    return result::ErrorExtensionNotSupported;
+                }
+            }
+        }
+
+        vk::ApplicationInfo appInfo{ desc.applicationName, VK_MAKE_VERSION(0, 0, 0), "Legion::LLRI", VK_MAKE_VERSION(0, 0, 1), VK_HEADER_VERSION_COMPLETE };
+        vk::InstanceCreateInfo ci{ {}, &appInfo, (uint32_t)layers.size(), layers.data(), (uint32_t)extensions.size(), extensions.data() };
+        ci.pNext = pNext;
+
+        vk::Instance vulkanInstance = nullptr;
+        const vk::Result createResult = vk::createInstance(&ci, nullptr, &vulkanInstance);
+
+        if (createResult != vk::Result::eSuccess)
+        {
+            llri::destroyInstance(result);
+            return internal::mapVkResult(createResult);
+        }
+
+        result->m_ptr = vulkanInstance;
+        *instance = result;
+        return result::Success;
     }
 
-    unsigned llri::Context::createSomeId(some_strategy_type creationStrategy) noexcept
+    void destroyInstance(Instance instance)
     {
-        return detail::VKStaticId + static_cast<unsigned>(creationStrategy);
+        if (!instance)
+            return;
+
+        for (auto& [ptr, adapter] : instance->m_cachedAdapters)
+            delete adapter;
+
+        //vk validation layers aren't tangible objects and don't need manual destruction
+
+        delete instance;
+    }
+
+    result InstanceT::enumerateAdapters(std::vector<Adapter>* adapters)
+    {
+        if (adapters == nullptr)
+            return result::ErrorInvalidUsage;
```

Comment:
If this was a reference you also wouldn't need to check if it was a nullptr. references can't be nullptrs and are always assumed to be altered. unless it's a const ref ofc.

Reply:
I don't think references indicate "out" parameters very clearly in an environment where we use const refs a lot, keeping out parameters as pointers is (I think) the preferred approach here